# **Section 1 — Creating & Reading Data**

In [1]:
import pandas as pd # Standard import

**Task 1:** 

Build a small DataFrame from a list of two dictionaries describing fish: the first
dictionary should have keys 'species', 'weight_kg', and 'price'; the second should have
only 'species' and 'weight_kg' (no price). Predict what the resulting DataFrame will look
like before you run it, then check.

There will be "NaN" appearing in the third column of the second row 

In [93]:
data_list = [{'species': 'shark', 'weight_kg': 700.0, 'price': 700}, {'species': 'gold fish', 'weight_kg': 0.2}]
df = pd.DataFrame(data_list)
df

,species,weight_kg,price
0,shark,700.0,700.0
1,gold fish,0.2,NaN


**Task 2:**

Import pandas and read abalone.csv into a DataFrame

In [10]:
import os
os.getcwd()

'c:\\Users\\micha\\Greenbootcamp\\DS_Python_advanced\\Assignments'

In [96]:
df = pd.read_csv("../data/abalone.csv")

# **Section 2 — First Look at Your Data**

**Task 3:**

Display the first 8 rows, then the last 8 rows.

In [97]:
df.head(8)

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
0,M,0.455,0.365,0.095,0.5140,0.2245,0.1010,0.150,15
1,M,0.350,0.265,0.090,0.2255,0.0995,0.0485,0.070,7
2,F,0.530,0.420,0.135,0.6770,0.2565,0.1415,0.210,9
3,M,0.440,0.365,0.125,0.5160,0.2155,0.1140,0.155,10
4,I,0.330,0.255,0.080,0.2050,0.0895,0.0395,0.055,7
5,I,0.425,0.300,0.095,0.3515,0.1410,0.0775,0.120,8
6,F,0.530,0.415,0.150,0.7775,0.2370,0.1415,0.330,20
7,F,0.545,0.425,0.125,0.7680,0.2940,0.1495,0.260,16


In [98]:
df.tail(8)

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
4169,M,0.520,0.385,0.165,0.7910,0.3750,0.1800,0.1815,10
4170,M,0.550,0.430,0.130,0.8395,0.3155,0.1955,0.2405,10
4171,M,0.560,0.430,0.155,0.8675,0.4000,0.1720,0.2290,8
4172,F,0.565,0.450,0.165,0.8870,0.3700,0.2390,0.2490,11
4173,M,0.590,0.440,0.135,0.9660,0.4390,0.2145,0.2605,10
4174,M,0.600,0.475,0.205,1.1760,0.5255,0.2875,0.3080,9
4175,F,0.625,0.485,0.150,1.0945,0.5310,0.2610,0.2960,10
4176,M,0.710,0.555,0.195,1.9485,0.9455,0.3765,0.4950,12


**Task 4:**

Check the shape and the column names.


In [99]:
df.shape

(4177, 9)

In [100]:
df.columns

Index(['Sex', 'Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight',
       'Viscera weight', 'Shell weight', 'Rings'],
      dtype='str')

**Task 5:**

Run info() and describe(). Based on the output alone, answer: how many total
rows are there, and what is the maximum value of the Height column?


In [101]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   str    
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole weight    4177 non-null   float64
 5   Shucked weight  4177 non-null   float64
 6   Viscera weight  4177 non-null   float64
 7   Shell weight    4177 non-null   float64
 8   Rings           4177 non-null   int64  
dtypes: float64(7), int64(1), str(1)
memory usage: 293.8 KB


"4177 entries" means 4177 rows

In [102]:
df.describe()

,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
count,4177.000000,4177.000000,4177.000000,4177.000000,4177.000000,4177.000000,4177.000000,4177.000000
mean,0.523992,0.407881,0.139516,0.828742,0.359367,0.180594,0.238831,9.933684
std,0.120093,0.099240,0.041827,0.490389,0.221963,0.109614,0.139203,3.224169
min,0.075000,0.055000,0.000000,0.002000,0.001000,0.000500,0.001500,1.000000
25%,0.450000,0.350000,0.115000,0.441500,0.186000,0.093500,0.130000,8.000000
50%,0.545000,0.425000,0.140000,0.799500,0.336000,0.171000,0.234000,9.000000
75%,0.615000,0.480000,0.165000,1.153000,0.502000,0.253000,0.329000,11.000000
max,0.815000,0.650000,1.130000,2.825500,1.488000,0.760000,1.005000,29.000000


1.13 

# **Section 3 — Getting Your Data**


**Task 6:**

Grab the 'Length' column two ways: bracket notation and dot notation. Confirm
both give the same result. Then try dot notation on the 'Shell weight' column — what
happens, and why?


In [103]:
df['Length']

0       0.455
1       0.350
2       0.530
3       0.440
4       0.330
        ...  
4172    0.565
4173    0.590
4174    0.600
4175    0.625
4176    0.710
Name: Length, Length: 4177, dtype: float64

In [104]:
df.Length

0       0.455
1       0.350
2       0.530
3       0.440
4       0.330
        ...  
4172    0.565
4173    0.590
4174    0.600
4175    0.625
4176    0.710
Name: Length, Length: 4177, dtype: float64

Same results for bracket and dot notation

In [22]:
df.Shell weight

SyntaxError: invalid syntax (1667055574.py, line 1)

An errors occurs, as the dot notation doesn't work with column names containing spaces

In [105]:
df['Shell weight']

0       0.1500
1       0.0700
2       0.2100
3       0.1550
4       0.0550
         ...  
4172    0.2490
4173    0.2605
4174    0.3080
4175    0.2960
4176    0.4950
Name: Shell weight, Length: 4177, dtype: float64

**Task 7:**

Use .loc[] to select rows 20 through 30, showing only the Sex and Diameter
columns.


In [ ]:
df.loc[20:30, ["Sex", "Diameter"]]        # index start and end are inclusive

,Sex,Diameter
20,M,0.280
21,I,0.275
22,F,0.440
23,F,0.415
24,F,0.480
25,F,0.440
26,F,0.450
27,M,0.445
28,M,0.475
29,M,0.425


**Task 8:**

Now get that exact same selection using .iloc[] instead, with integer positions

In [ ]:
df.iloc[20:31, [0, 2]]      # index end is exclusive, column index e.g. from df.info

,Sex,Diameter
20,M,0.280
21,I,0.275
22,F,0.440
23,F,0.415
24,F,0.480
25,F,0.440
26,F,0.450
27,M,0.445
28,M,0.475
29,M,0.425


**Task 9:**

Build a mask for rows where Diameter is greater than 0.5, and use it to filter the
DataFrame. Then add a second condition — Sex equal to 'M' — to that same mask.

In [29]:
df["Diameter"] > 0.5
df[df["Diameter"] > 0.5]

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
31,F,0.680,0.560,0.165,1.6390,0.6055,0.2805,0.460,15
32,M,0.665,0.525,0.165,1.3380,0.5515,0.3575,0.350,18
33,F,0.680,0.550,0.175,1.7980,0.8150,0.3925,0.455,19
34,F,0.705,0.550,0.200,1.7095,0.6330,0.4115,0.490,13
81,M,0.620,0.510,0.175,1.6150,0.5105,0.1920,0.675,12
...,...,...,...,...,...,...,...,...,...
4145,M,0.670,0.525,0.200,1.7405,0.6205,0.2970,0.657,11
4146,M,0.695,0.530,0.210,1.5100,0.6640,0.4095,0.385,10
4147,M,0.695,0.550,0.195,1.6645,0.7270,0.3600,0.445,11
4148,M,0.770,0.605,0.175,2.0505,0.8005,0.5260,0.355,11


In [30]:
df[(df["Diameter"] > 0.5) & (df["Sex"] == "M")]

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
32,M,0.665,0.525,0.165,1.3380,0.5515,0.3575,0.350,18
81,M,0.620,0.510,0.175,1.6150,0.5105,0.1920,0.675,12
94,M,0.695,0.560,0.190,1.4940,0.5880,0.3425,0.485,15
95,M,0.665,0.535,0.195,1.6060,0.5755,0.3880,0.480,14
128,M,0.700,0.535,0.160,1.7255,0.6300,0.2635,0.540,19
...,...,...,...,...,...,...,...,...,...
4145,M,0.670,0.525,0.200,1.7405,0.6205,0.2970,0.657,11
4146,M,0.695,0.530,0.210,1.5100,0.6640,0.4095,0.385,10
4147,M,0.695,0.550,0.195,1.6645,0.7270,0.3600,0.445,11
4148,M,0.770,0.605,0.175,2.0505,0.8005,0.5260,0.355,11


**Task 10:**

Rewrite the compound filter from Task 9 using .query() instead of bracket-andmask syntax. (Hint: you'll need to handle the space in 'Whole weight'-style column
names if you reference any of them — stick to Diameter and Sex and you won't need
to.)

In [ ]:
df.query("Diameter > 0.5 and Sex == 'M'")       # without apostrophe around M the code won't work

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
32,M,0.665,0.525,0.165,1.3380,0.5515,0.3575,0.350,18
81,M,0.620,0.510,0.175,1.6150,0.5105,0.1920,0.675,12
94,M,0.695,0.560,0.190,1.4940,0.5880,0.3425,0.485,15
95,M,0.665,0.535,0.195,1.6060,0.5755,0.3880,0.480,14
128,M,0.700,0.535,0.160,1.7255,0.6300,0.2635,0.540,19
...,...,...,...,...,...,...,...,...,...
4145,M,0.670,0.525,0.200,1.7405,0.6205,0.2970,0.657,11
4146,M,0.695,0.530,0.210,1.5100,0.6640,0.4095,0.385,10
4147,M,0.695,0.550,0.195,1.6645,0.7270,0.3600,0.445,11
4148,M,0.770,0.605,0.175,2.0505,0.8005,0.5260,0.355,11


# **Section 4 — Groupby & Sorting**

**Task 11:**

Group the data by Sex and find the mean Shell weight for each group.


In [49]:
df.groupby("Sex").mean()["Shell weight"]

Sex
F    0.302010
I    0.128182
M    0.281969
Name: Shell weight, dtype: float64

**Task 12:**

Sort the DataFrame by Whole weight in descending order and show the top
10 rows.

In [51]:
df.sort_values("Whole weight", ascending=False).head(10)

,Sex,Length,Diameter,Height,Whole weight,Shucked weight,Viscera weight,Shell weight,Rings
891,M,0.730,0.595,0.230,2.8255,1.1465,0.4190,0.8970,17
1763,M,0.775,0.630,0.250,2.7795,1.3485,0.7600,0.5780,12
1209,F,0.780,0.630,0.215,2.6570,1.4880,0.4985,0.5860,11
1051,F,0.735,0.600,0.220,2.5550,1.1335,0.4400,0.6000,11
165,M,0.725,0.570,0.190,2.5500,1.0705,0.4830,0.7250,14
3715,M,0.780,0.600,0.210,2.5480,1.1945,0.5745,0.6745,11
2334,F,0.800,0.630,0.195,2.5260,0.9330,0.5900,0.6200,23
1762,M,0.770,0.620,0.195,2.5155,1.1155,0.6415,0.6420,12
1427,F,0.750,0.610,0.235,2.5085,1.2320,0.5190,0.6120,14
1207,F,0.755,0.625,0.210,2.5050,1.1965,0.5130,0.6785,11


# **Section 5 — Creating & Dropping Columns**

**Task 13:**

Rename the Rings column to age_proxy. Then create a new column called
density_proxy that holds Whole weight divided by Length, using .eval().


In [90]:
df = pd.read_csv("../data/abalone.csv") 
# necessary, because wrong code messed up my df and "Rings" was gone                                                                          

df.rename(columns={"Rings": "age_proxy", "Whole weight": "Whole_weight"}, inplace=True)
df.eval("density_proxy = Whole_weight / Length", inplace=True)
df.info()


<class 'pandas.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   str    
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole_weight    4177 non-null   float64
 5   Shucked weight  4177 non-null   float64
 6   Viscera weight  4177 non-null   float64
 7   Shell weight    4177 non-null   float64
 8   age_proxy       4177 non-null   int64  
 9   density_proxy   4177 non-null   float64
dtypes: float64(8), int64(1), str(1)
memory usage: 326.5 KB


**Task 14:**

Drop age_proxy properly, so the change actually sticks. Confirm it's gone.
Then run fillna(-1) and dropna() on the DataFrame — even though nothing will visibly
change, since there's no missing data here, this is just to practice the syntax.

In [91]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   str    
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole_weight    4177 non-null   float64
 5   Shucked weight  4177 non-null   float64
 6   Viscera weight  4177 non-null   float64
 7   Shell weight    4177 non-null   float64
 8   age_proxy       4177 non-null   int64  
 9   density_proxy   4177 non-null   float64
dtypes: float64(8), int64(1), str(1)
memory usage: 326.5 KB


In [92]:
df.drop("age_proxy", inplace=True, axis=1)
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4177 entries, 0 to 4176
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Sex             4177 non-null   str    
 1   Length          4177 non-null   float64
 2   Diameter        4177 non-null   float64
 3   Height          4177 non-null   float64
 4   Whole_weight    4177 non-null   float64
 5   Shucked weight  4177 non-null   float64
 6   Viscera weight  4177 non-null   float64
 7   Shell weight    4177 non-null   float64
 8   density_proxy   4177 non-null   float64
dtypes: float64(8), str(1)
memory usage: 293.8 KB


In [88]:
df.fillna(-1)

,Sex,Length,Diameter,Height,Whole_weight,Shucked weight,Viscera weight,Shell weight
0,M,0.455,0.365,0.095,0.5140,0.2245,0.1010,0.1500
1,M,0.350,0.265,0.090,0.2255,0.0995,0.0485,0.0700
2,F,0.530,0.420,0.135,0.6770,0.2565,0.1415,0.2100
3,M,0.440,0.365,0.125,0.5160,0.2155,0.1140,0.1550
4,I,0.330,0.255,0.080,0.2050,0.0895,0.0395,0.0550
...,...,...,...,...,...,...,...,...
4172,F,0.565,0.450,0.165,0.8870,0.3700,0.2390,0.2490
4173,M,0.590,0.440,0.135,0.9660,0.4390,0.2145,0.2605
4174,M,0.600,0.475,0.205,1.1760,0.5255,0.2875,0.3080
4175,F,0.625,0.485,0.150,1.0945,0.5310,0.2610,0.2960


In [89]:
df.dropna()

,Sex,Length,Diameter,Height,Whole_weight,Shucked weight,Viscera weight,Shell weight
0,M,0.455,0.365,0.095,0.5140,0.2245,0.1010,0.1500
1,M,0.350,0.265,0.090,0.2255,0.0995,0.0485,0.0700
2,F,0.530,0.420,0.135,0.6770,0.2565,0.1415,0.2100
3,M,0.440,0.365,0.125,0.5160,0.2155,0.1140,0.1550
4,I,0.330,0.255,0.080,0.2050,0.0895,0.0395,0.0550
...,...,...,...,...,...,...,...,...
4172,F,0.565,0.450,0.165,0.8870,0.3700,0.2390,0.2490
4173,M,0.590,0.440,0.135,0.9660,0.4390,0.2145,0.2605
4174,M,0.600,0.475,0.205,1.1760,0.5255,0.2875,0.3080
4175,F,0.625,0.485,0.150,1.0945,0.5310,0.2610,0.2960
